relay-replays/src/recording/transcoder.rs

```diff
@@ -0,0 +1,310 @@
+//! Serde transcoder modeled after `serde_transcode`.
+
+use std::cell::RefCell;
+use std::fmt;
+
+use serde::de;
+use serde::ser::{self, SerializeMap, SerializeSeq};
+
+/// A Serde transcoder that transforms strings.
+///
+/// Streams all contents from the deserializer into the serializer and passes
+/// strings through a mutation callback.
+///
+/// # Note
+///
+/// Unlike traditional serializable types, `StringTranscoder`'s `Serialize`
+/// implementation is *not* idempotent, as it advances the state of its
+/// internal `Deserializer`. It should only ever be serialized once.
+pub struct StringTranscoder<'a, D>(RefCell<Option<D>>, &'a dyn Fn(&mut String));
+
+impl<'de, 'a, D> StringTranscoder<'a, D>
+where
+    D: de::Deserializer<'de>,
+{
+    /// Constructs a new `StringTranscoder`.
+    pub fn new(d: D, transform: &'a dyn Fn(&mut String)) -> StringTranscoder<D> {
+        StringTranscoder(RefCell::new(Some(d)), transform)
+    }
+}
+
+impl<'de, 'a, D> ser::Serialize for StringTranscoder<'a, D>
+where
+    D: de::Deserializer<'de>,
+{
+    fn serialize<S>(&self, s: S) -> Result<S::Ok, S::Error>
+    where
+        S: ser::Serializer,
+    {
+        let Self(ref d, ref f) = self;
+
+        d.borrow_mut()
+            .take()
+            .expect("Transcoder may only be serialized once")
+            .deserialize_any(Visitor(s, f))
+            .map_err(d2s)
+    }
+}
+
+struct Visitor<'a, S>(S, &'a dyn Fn(&mut String));
```

Comment:
nit: Give names to fields?

Reply:
Good idea, I'll give the transcoder a refactor in a follow-up.